packages/core/data-transfer/lib/bootstrap/controllers/transfer.ts

```diff
@@ -0,0 +1,147 @@
+import createPushController from './push';
+
+interface ITransferState {
+  kind?: TransferKind;
+  transferID?: string;
+  controller?: IPushController;
+}
+
+const createTransferController =
+  (options: ServerOptions = {}) =>
+  async (ctx: Context) => {
+    const upgradeHeader = (ctx.request.headers.upgrade || '')
+      .split(',')
+      .map((s) => s.trim().toLowerCase());
+
+    // Create the websocket server
+    const wss = new WebSocket.Server({ ...options, noServer: true });
+
+    if (upgradeHeader.includes('websocket')) {
+      wss.handleUpgrade(ctx.req, ctx.request.socket, Buffer.alloc(0), (ws) => {
+        // Create a connection between the client & the server
+        wss.emit('connection', ws, ctx.req);
+
+        const state: ITransferState = {};
+        let uuid: string | undefined;
+
+        const callback = <T = unknown>(e: Error | null = null, data?: T) => {
+          return new Promise<void>((resolve, reject) => {
+            if (!uuid) {
+              reject(new Error('Missing uuid for this message'));
+              return;
+            }
+
+            const payload = JSON.stringify({
+              uuid,
+              data: data ?? {},
+              error: e,
+            });
+
+            ws.send(payload, (error) => (error ? reject(error) : resolve()));
+          });
+        };
+
+        const answer = async <T = unknown>(fn: () => T) => {
+          try {
+            const response = await fn();
+            callback(null, response);
+          } catch (e) {
+            if (e instanceof Error) {
+              callback(e);
+            } else if (typeof e === 'string') {
+              callback(new Error(e));
+            } else {
+              callback(new Error('Unexpected error'));
+            }
+          }
+        };
+
+        const teardown = () => {
+          delete state.kind;
+          delete state.controller;
+          delete state.transferID;
+
+          return { ok: true };
+        };
+
+        const init = (msg: InitMessage) => {
+          const { kind, options: controllerOptions } = msg;
+
+          if (state.controller) {
+            throw new Error('Transfer already in progres');
+          }
+
+          if (kind === 'push') {
+            state.controller = createPushController({
+              ...controllerOptions,
+              autoDestroy: false,
+              getStrapi() {
+                return strapi;
+              },
+            });
+          }
+
+          // Pull or others
+          else {
+            throw new Error(`${kind} transfer not implemented`);
+          }
+
+          state.transferID = v4();
+
+          return { transferID: state.transferID };
+        };
+
+        ws.on('close', () => {
+          teardown();
+        });
+
+        ws.on('error', (e) => {
+          teardown();
+          console.error(e);
+        });
+
+        ws.on('message', async (raw) => {
+          const msg: Message = JSON.parse(raw.toString());
```

Comment:
JSON.parse is prone to throwing errors, do we want to wrap this in a try/catch and handle it on our own terms? otherwise I think this will get an "error thrown in an async method" if we get a bad input message

Reply:
I think that might be what was happening with your change in the way errors where sent, what do you think?

Reply:
Ahh, I see, so we do handle them there. This goes with the comment I had elsewhere about how we should try to have some standard error codes for what we return from the websocket, so that (in this example) instead of getting an error directly from JSON.parse we send something like "Invalid JSON in the request"